fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/DynamicRegistriesImpl.java

```diff
@@ -50,16 +53,78 @@ private DynamicRegistriesImpl() {
 	}
 
 	public static @Unmodifiable List<RegistryLoader.Entry<?>> getDynamicRegistries() {
-		if (!sorted) {
-			sort();
-			sorted = true;
+		List<RegistryLoader.Entry<?>> ret = sortedRegistries;
+
+		if (ret == null) {
+			sortedRegistries = ret = sort();
+		}
+
+		return ret;
+	}
+
+	private static List<RegistryLoader.Entry<?>> sort() {
+		Map<RegistryKey<? extends Registry<?>>, RegistryNode> nodes = new HashMap<>(RegistryLoader.DYNAMIC_REGISTRIES.size() + SETTINGS.size());
+
+		// Add vanilla nodes with their ordering
+		int vanillaIndex = 0;
+		RegistryNode previousVanillaNode = null;
+
+		for (RegistryLoader.Entry<?> vanillaEntry : RegistryLoader.DYNAMIC_REGISTRIES) {
+			RegistryNode node = new RegistryNode(vanillaEntry, vanillaIndex++);
+			nodes.put(vanillaEntry.key(), node);
+
+			if (previousVanillaNode != null) {
+				link(previousVanillaNode, node);
+			}
+
+			previousVanillaNode = node;
+		}
+
+		// Add modded nodes
+		for (SettingsImpl<?> settings : SETTINGS.values()) {
+			RegistryNode node = nodes.computeIfAbsent(settings.owner.key(), k -> new RegistryNode(settings.owner, RegistryNode.MODDED_INDEX));
+			nodes.put(settings.owner.key(), node);
+		}
+
+		// Add modded ordering
+		for (SettingsImpl<?> settings : SETTINGS.values()) {
+			RegistryNode node = nodes.get(settings.owner.key());
+
+			for (RegistryKey<? extends Registry<?>> before : settings.before) {
+				RegistryNode other = nodes.get(before);
+
+				if (other == null) {
+					throw new IllegalStateException("Registry " + settings.owner.key() + " has a dependency on " + before + ", which does not exist!");
+				}
+
+				link(node, other);
+			}
+
+			for (RegistryKey<? extends Registry<?>> after : settings.after) {
+				RegistryNode other = nodes.get(after);
+
+				if (other == null) {
+					throw new IllegalStateException("Registry " + settings.owner.key() + " has a dependency on " + after + ", which does not exist!");
+				}
+
+				link(other, node);
+			}
+		}
+
+		// Sort everything
+		List<RegistryNode> nodesToSort = new ArrayList<>(nodes.values());
+		NodeSorting.sort(nodesToSort, "dynamic registries", RegistryNode.COMPARATOR);
+
+		for (RegistryNode node : nodesToSort) {
+			System.out.println("Sorted node: " + node.entry.key());
 		}
 
-		return List.copyOf(DYNAMIC_REGISTRIES);
+		return nodesToSort.stream().map(node -> node.entry).collect(Collectors.toUnmodifiableList());
 	}
 
-	private static void sort() {
-		// TODO
+	private static void link(RegistryNode node1, RegistryNode node2) {
+		node1.subsequentNodes.add(node2);
+		node2.previousNodes.add(node1);
 	}
```

Comment:
This feels like something that belongs on `SortableNode`? Some method like `SortableNode<N>.linkTo(N after)`

Reply:
Probably as a static method